// scanpkg/src/scanpkg.js
/* 
* @Author: mustafa
* @Date:   2015-07-30 22:06:41
* @Last Modified by:   mustafa
* @Last Modified time: 2015-08-01 11:01:52
*/

var fs = require("fs"),
	path = require("path"),
	child_process = require("child_process"),
	checksum = require("checksum"),
	_control = require("./control.js");

var noDirError = "scanpkgNoDir",
	emptyDirError = "scanpkgEmptyDir";

exports.noDirError = noDirError;
exports.emptyDirError = emptyDirError;

// Fix windows path separators
path.join2 = path.join;
path.sep = '/';
path.join = function(){
    var res = path.join2.apply({}, arguments);
    res = res.replace(/\\/g, path.sep);
    return res;
}

function getFilesizeInBytes(filename) {
	var stats = fs.statSync(filename);
	return stats["size"];
}

function modifyControl(control, deb, dir) {
	var debContent = fs.readFileSync(deb),
		md5 = checksum(debContent, {"algorithm": "md5"}),
		sha1 = checksum(debContent, {"algorithm": "sha1"}),
		sha256 = checksum(debContent, {"algorithm": "sha256"}),
		filename = path.join(path.basename(dir), path.basename(deb)),
		fileSize = getFilesizeInBytes(deb);

	control["MD5sum"] = md5;
	control["SHA1"] = sha1;
	control["SHA256"] = sha256;
	control["Filename"] = filename;
	control["Size"] = fileSize;

	return _control.toControl(control);
}

exports.new = function(dir, callback) {
    console.log("Scanning packages...");

	var tmpDir = path.join(dir, ".scanpkg"),
		returnString = "";

	if (fs.existsSync(tmpDir)) {
		child_process.spawnSync("rm", ["-rf", tmpDir]);
	}
	fs.mkdirSync(tmpDir);

	if (typeof(callback) != "function") {
		console.error("You have not provided a valid callback to scanpkg.new");
		return;
	}

	fs.lstat(dir, function(err, stats) {
		if (err || !stats.isDirectory())
			return callback(noDirError);

		var filesInDir = fs.readdirSync(dir);
		if (filesInDir.length < 1)
			return callback(emptyDirError);
		
		// Iterate through files in dir
		for (var i = 0; i < filesInDir.length; i++) {
			var filePath = path.join(dir, filesInDir[i]);
			if (filesInDir[i].indexOf(".DS_Store") != -1) {
				fs.unlinkSync(filePath);
			} else if (path.extname(filePath) == ".deb") {
				var tempPackageDir = path.join(tmpDir, path.basename(filePath, ".deb")),
					tempPackageControlTar = path.join(tempPackageDir, "control.tar.gz"),
					tempPackageControl = path.join(tempPackageDir, "control");
				
				fs.mkdirSync(tempPackageDir);
				child_process.spawnSync("cp", [filePath, tempPackageDir]);
				filePath = path.join(tempPackageDir, filesInDir[i]);
				child_process.spawnSync("ar", ["-x", filePath]);
				child_process.spawnSync("rm", ["data.tar.gz", "debian-binary"]);
				child_process.spawnSync("rm", ["data.tar", "debian-binary"]);
				child_process.spawnSync("mv", ["control.tar.gz", tempPackageDir]);
				child_process.spawnSync("tar", ["-xvzf", tempPackageControlTar, "-C", tempPackageDir]);

				var __control = _control.new(fs.readFileSync(tempPackageControl).toString("utf8")),
					modified_control = modifyControl(__control, filePath, dir);

				returnString += modified_control + "\n";
			}
		}

		child_process.spawnSync("rm", ["-rf", tmpDir]);

		if(returnString.lastIndexOf("\n") > 0) {
			return callback(null, returnString.substring(0, returnString.lastIndexOf("\n")));
		} else {
			return callback(null, returnString);
		}
	});
}
